text/0000-option-unwrap-unchecked.md

```diff
@@ -0,0 +1,73 @@
+- Feature Name: option_unwrap_unchecked
+- Start Date: 2015-4-28
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC adds a method to the `Option` enum which allows it to be
+unsafely unwrapped without checking for `None`.
+
+# Motivation
+
+In cases where a developer can determine that an `Option` is `Some(v)`
+but the compiler cannot, this method allows a developer to unsafely
+opt-in to an optimization which will avoid a conditional check. In my
+own uses I find cases where I know that an `Option` is `Some(v)`. I
+present a case from the `libcollections linked_list.rs`:
+
+    fn next(&mut self) -> Option<&'a A> {
+        if self.nelem == 0 {
+            return None;
+        }
+        self.head.as_ref().map(|head| {
+            self.nelem -= 1;
+            self.head = &head.next;
+            &head.value
+        })
+    }
+
+The check for `nelem == 0` already ensures that `self.head.as_ref()`
+cannot be `None` but the `map` method is unlikely to optimize out the
+check. Here, an unsafe method which could unwrap the `Option` would
+allow the compiler to optimize. I expect that this can be used in many
+more cases throughout both the standard library and other code. The
+outcome is likely a minor performance improvement, but one that should
+be possible for developers to make.
+
+# Detailed design
+
+I choose to name this method `unwrap_unchecked()` (commence
+bike-shedding). It is implemented quite simply by using the
+`unreachable` intrinsic to indicate that the `Option` cannot be `None`
+and then calling `unwrap()`. I expect that the compiler can them
+optimize out the conditional. A prototype implement can be found
```

Comment:
This is dependent on a lot of different cases. If I build a library as follows:

    #![feature(core)]

    use std::intrinsics;

    pub fn my_unwrap(o: Option<i32>) -> i32 {
        match o {
            Some(v) => v,
            None => panic!("panic")
        }
    }
    
    pub unsafe fn my_unwrap_unsafe(o: Option<i32>) -> i32 {
        match o {
            Some(v) => v,
            None => intrinsics::unreachable()
        }
    }


    $ rustc -O --crate-type=lib unwrap.rs
    $ objdump -d libunwrap.rlib
    0000000000000000 <_ZN16my_unwrap_unsafe20h8f4545557bc4e296MaaE>:
       0:	48 c1 ef 20          	shr    $0x20,%rdi
       4:	89 f8                	mov    %edi,%eax
       6:	c3                   	retq   


You can see no conditionals in the unwrap_unsafe asm. Further testing on the actual standard library implementation would be needed to confirm, but at least one method of implementing it seems to work as I expect.